context.go

```diff
@@ -409,40 +411,41 @@ func (m *Macaron) SetDefaultCookieSecret(secret string) {
 
 // SetSecureCookie sets given cookie value to response header with default secret string.
 func (ctx *Context) SetSecureCookie(name, value string, others ...interface{}) {
-	ctx.SetSuperSecureCookie(defaultCookieSecret, name, value, others...)
+	ctx.SetSuperSecureCookie([]byte(defaultCookieSecret), name, value, others...)
 }
 
 // GetSecureCookie returns given cookie value from request header with default secret string.
 func (ctx *Context) GetSecureCookie(key string) (string, bool) {
-	return ctx.GetSuperSecureCookie(defaultCookieSecret, key)
+	return ctx.GetSuperSecureCookie([]byte(defaultCookieSecret), key)
 }
 
 // SetSuperSecureCookie sets given cookie value to response header with secret string.
-func (ctx *Context) SetSuperSecureCookie(secret, name, value string, others ...interface{}) {
-	m := md5.Sum([]byte(secret))
-	secret = hex.EncodeToString(m[:])
-	text, err := com.AESEncrypt([]byte(secret), []byte(value))
+func (ctx *Context) SetSuperSecureCookie(password []byte, name, value string, others ...interface{}) []byte {
+	salt := make([]byte, 128)
+	if _, err := rand.Read(salt); err != nil {
+		panic("Error generating 128-bit salt")
+	}
+
+	key := pbkdf2.Key(password, salt, 10000, 32, sha256.New)
+	text, err := com.AESGCMEncrypt(key, []byte(value))
 	if err != nil {
 		panic("error encrypting cookie: " + err.Error())
 	}
+
 	ctx.SetCookie(name, hex.EncodeToString(text), others...)
+
+	return key
```

Comment:
The PBKDF2 stuff is just for generating a key to be used with AES. So when you pass a "secret" from Gogs for example that value is being used to derive a key which will then be used for AES encryption/decryption.

Before these changes Gogs was still passing a "secret" over but it was then hashed/hex encoded and used as a key to AES. Because PBKDF2 uses a salt the key will actually be different every time, even if you pass in a "secret" with the value of `userpassword+token` or whatever it was in Gogs.